diff --git a/arch/arm/boot/dts/zynq-enclustra-qspi-parts.dtsi b/arch/arm/boot/dts/zynq-enclustra-qspi-parts.dtsi
index 9df2a7ec3..fe0ea91e2 100644
--- a/arch/arm/boot/dts/zynq-enclustra-qspi-parts.dtsi
+++ b/arch/arm/boot/dts/zynq-enclustra-qspi-parts.dtsi
@@ -9,36 +9,58 @@
 	is-dual = <0>;
 	num-cs = <1>;
 	flash@0 {
-		compatible = "n25q128a11";
+		compatible = "s25fl512s";
 		reg = <0x0>;
 		spi-tx-bus-width = <1>;
 		spi-rx-bus-width = <4>;
 		spi-max-frequency = <50000000>;
 		#address-cells = <1>;
 		#size-cells = <1>;
-		partition@qspi-bootimage {
-			label = "qspi-bootimage";
-			reg = <0x0 0x600000>;
+		partition@0 {
+			label = "boot";
+			reg = <0x0 0x500000>;
+      read-only;
 		};
-		partition@qspi-kernel {
-			label = "qspi-kernel";
-			reg = <0x600000 0x500000>;
+    partition@500000 {
+      label = "boot-script";
+      reg = <0x500000 0x40000>;
+      read-only;
+    };
+    partition@540000 {
+      label = "boot-env";
+      reg = <0x540000 0x80000>;
+    };
+    partition@5C0000 {
+      label = "kernel-a";
+      reg = <0x5C0000 0xF80000>;
+    };
+    partition@A00000 {
+      label = "devicetree-a";
+      reg = <0xA00000 0x40000>;
+    };
+    partition@A40000 {
+      label = "rootfs-a";
+      reg = <0xA40000 0xB00000>;
+    };
+		partition@1540000 {
+			label = "kernel-b";
+			reg = <0x1540000 0xF80000>;
 		};
-		partition@qspi-device-tree {
-			label = "qspi-device-tree";
-			reg = <0xB00000 0x80000>;
+		partition@1980000 {
+			label = "devicetree-b";
+			reg = <0x1980000 0x40000>;
 		};
-		partition@qspi-bootargs {
-			label = "qspi-bootargs";
-			reg = <0xB80000 0x80000>;
+		partition@19C0000 {
+			label = "rootfs-b";
+			reg = <0x19C0000 0xB00000>;
 		};
-		partition@qspi-bootscript {
-			label = "qspi-bootscript";
-			reg = <0xC00000 0x40000>;
+		partition@24C0000 {
+			label = "rootfs-data";
+			reg = <0x24C0000 0x140000>;
 		};
-		partition@qspi-rootfs {
-			label = "qspi-rootfs";
-			reg = <0xC40000 0x3C0000>;
+		partition@25C0000 {
+			label = "appfs";
+			reg = <0x25C0000 0x1A00000>;
 		};
 	};
 };
diff --git a/arch/arm/boot/dts/zynq-enclustra-zx-common.dtsi b/arch/arm/boot/dts/zynq-enclustra-zx-common.dtsi
index 796556847..d945bf80c 100644
--- a/arch/arm/boot/dts/zynq-enclustra-zx-common.dtsi
+++ b/arch/arm/boot/dts/zynq-enclustra-zx-common.dtsi
@@ -1,32 +1,32 @@
 / {
-	memory {
+	memory@0 {
 		device_type = "memory";
 		reg = <0x0 0x20000000>;
 	};
 
 	chosen {
-		linux,stdout-path = "/amba/serial@e0000000";
+		stdout-path = &uart1;
 	};
 
-	usb_phy0: phy0 {
+	usb_phy0: phy0@e0002000 {
 		compatible = "ulpi-phy";
 		#phy-cells = <0>;
 		reg = <0xe0002000 0x1000>;
 		view-port = <0x0170>;
 		drv-vbus;
 	};
-	mdio {
-                compatible = "cdns,macb-mdio";
-                reg = <0xe000b000 0x1000>;
-                clocks = <&clkc 30>, <&clkc 30>, <&clkc 13>;
-                clock-names = "pclk", "hclk", "tx_clk";
-                #address-cells = <1>;
-                #size-cells = <0>;
-                phy3: ethernet-phy@3 {
-                        reg = <3>;
-                };
-        };
 
+	mdio@e000b000 {
+		compatible = "cdns,macb-mdio";
+		reg = <0xe000b000 0x1000>;
+		clocks = <&clkc 30>, <&clkc 30>, <&clkc 13>;
+		clock-names = "pclk", "hclk", "tx_clk";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		phy3: ethernet-phy@3 {
+		        reg = <3>;
+		};
+	};
 };
 
 &gem0 {
diff --git a/arch/arm/boot/dts/zynq-mars-zx3.dts b/arch/arm/boot/dts/zynq-mars-zx3.dts
index ffe11dc91..48e8f7352 100644
--- a/arch/arm/boot/dts/zynq-mars-zx3.dts
+++ b/arch/arm/boot/dts/zynq-mars-zx3.dts
@@ -15,6 +15,7 @@
 /include/ "zynq-7000.dtsi"
 /include/ "zynq-enclustra-zx-common.dtsi"
 /include/ "zynq-enclustra-qspi-parts.dtsi"
+/include/ "zynq-tsat-3500-parts.dtsi"
 
 / {
 	model = "Enclustra Mars ZX3 SOM";
diff --git a/arch/arm/boot/dts/zynq-tsat-3500-parts.dtsi b/arch/arm/boot/dts/zynq-tsat-3500-parts.dtsi
new file mode 100644
index 000000000..dfce37684
--- /dev/null
+++ b/arch/arm/boot/dts/zynq-tsat-3500-parts.dtsi
@@ -0,0 +1,220 @@
+/*
+ *  TSAT 3500
+ */
+
+/ {
+	amba: amba {
+		uart2: serial@43c10000 {
+			clock-frequency = <100000000>;
+			clock-names = "s_axi_aclk";
+			clocks = <&clkc 15>;
+			compatible = "xlnx,xps-uart16550", "ns16550a";
+			status = "disabled";
+			current-speed = <115200>;
+			interrupt-names = "ip2intc_irpt";
+			interrupt-parent = <&intc>;
+			interrupts = <0 33 4>;
+			port-number = <0>;
+			reg = <0x43c10000 0x10000>;
+			reg-offset = <0x1000>;
+			reg-shift = <2>;
+			xlnx,external-xin-clk-hz = <0x17d7840>;
+			xlnx,external-xin-clk-hz-d = <0x19>;
+			xlnx,has-external-rclk = <0x0>;
+			xlnx,has-external-xin = <0x0>;
+			xlnx,is-a-16550 = <0x1>;
+			xlnx,s-axi-aclk-freq-hz-d = "100.0";
+			xlnx,use-modem-ports = <0x1>;
+			xlnx,use-user-ports = <0x1>;
+		};
+	};
+
+	aliases {
+		ethernet0 = &gem0;
+		ethernet1 = &gem1;
+		mmc0 = &sdhci0;
+		serial0 = &uart0;
+		serial1 = &uart1;
+		serial2 = &uart2;
+		spi0 = &qspi;
+		usb0 = &usb0;
+		gpio0 = &gpio0;
+		gpio1 = &tca6416_led;
+		gpio2 = &tca6416_gpio;
+		i2c0 = &i2c0;
+		i2c1 = &i2c1;
+	};
+
+	mdio1@e000c000 {
+    compatible = "cdns,macb-mdio";
+    reg = <0xe000c000 0x1000>;
+    clocks = <&clkc 31>, <&clkc 31>;
+		clock-names = "pclk", "hclk";
+    #address-cells = <1>;
+    #size-cells = <0>;
+		phy7: ethernet-phy@7 {
+			reg = <7>;
+		};
+	};
+
+	fpga_reg_fft@44000000 {
+		compatible = "generic-uio";
+		reg = <0x44000000 0x1000>;
+		interrupt-parent = <0x4>;
+		interrupts = <0x0 0x1d 0x1>;
+	};
+
+	fpga_reg_data@44000000 {
+		compatible = "generic-uio";
+		reg = <0x44000000 0x1000>;
+		interrupt-parent = <0x4>;
+		interrupts = <0x0 0x1e 0x1>;
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+		led0 {
+			label = "board:green:app:tx";
+			gpios = <&tca6416_led 0 0>;
+			default-state = "off";
+		};
+		led1 {
+			label = "board:red:app:tx";
+			gpios = <&tca6416_led 1 0>;
+			default-state = "off";
+		};
+		led2 {
+			label = "board:green:app:rx";
+			gpios = <&tca6416_led 2 0>;
+			default-state = "off";
+		};
+		led3 {
+			label = "board:red:app:rx";
+			gpios = <&tca6416_led 3 0>;
+			default-state = "off";
+		};
+		led4 {
+			label = "board:green:sat:tx";
+			gpios = <&tca6416_led 4 0>;
+			default-state = "off";
+		};
+		led5 {
+			label = "board:red:sat:tx";
+			gpios = <&tca6416_led 5 0>;
+			default-state = "off";
+		};
+		led6 {
+			label = "board:green:sat:rx";
+			gpios = <&tca6416_led 6 0>;
+			default-state = "off";
+		};
+		led7 {
+			label = "board:red:sat:rx";
+			gpios = <&tca6416_led 7 0>;
+			default-state = "off";
+		};
+		led8 {
+			label = "board:green:lnk";
+			gpios = <&tca6416_led 8 0>;
+			default-state = "off";
+		};
+		led9 {
+			label = "board:red:lnk";
+			gpios = <&tca6416_led 9 0>;
+			default-state = "off";
+		};
+		led10 {
+			label = "board:green:loc";
+			gpios = <&tca6416_led 10 0>;
+			default-state = "off";
+		};
+		led11 {
+			label = "board:red:loc";
+			gpios = <&tca6416_led 11 0>;
+			default-state = "off";
+		};
+		led12 {
+			label = "board:green:sys";
+			gpios = <&tca6416_led 12 0>;
+			default-state = "off";
+		};
+		led13 {
+			label = "board:red:sys";
+			gpios = <&tca6416_led 13 0>;
+			default-state = "off";
+		};
+	};
+};
+
+&uart1 {
+	status = "okay";
+};
+
+&uart2 {
+	status = "okay";
+};
+
+&gem1 {
+	status = "okay";
+	phy-mode = "mii-id";
+	phy-handle = <&phy7>;
+	xlnx,emio = <1>;
+};
+
+&i2c0 {
+	status = "okay";
+	zynq_eeprom: eeprom@5c {
+		compatible = "24c01"; // Maxim DS28CN01
+		reg = <0x5c>;
+		read-only;
+	};
+};
+
+&i2c1 {
+	status = "okay";
+
+	tca6416_led: gpio@20 {
+		compatible = "ti,tca6416";
+		reg = <0x20>;
+		gpio-controller;
+		#gpio-cells = <2>;
+	};
+
+	tca6416_gpio: gpio@21 {
+		compatible = "ti,tca6416";
+		reg = <0x21>;
+		gpio-controller;
+		#gpio-cells = <2>;
+	};
+
+	ina226_rx: ina226@40 {
+		compatible = "ti,ina226";
+		ti,rail-name = "RX";
+		reg = <0x40>;
+		shunt-resistor = <40000>;
+	};
+
+	ina226_tx: ina226@41 {
+		compatible = "ti,ina226";
+		ti,rail-name = "TX";
+		reg = <0x41>;
+		shunt-resistor = <10000>;
+	};
+/*
+	ADS1000@48 {
+
+	}
+*/
+	eeprom: eeprom@50 {
+		compatible = "atmel,24c04";
+		reg = <0x50>;
+	};
+
+	adrf6755: adrf6755@60 {
+		status = "okay";
+		compatible = "adrf6755";
+		reg = <0x60>;
+		ref_frequency_KHz = <0x2710>;
+		pfd_frequency_KHz = <0x4e20>;
+	};
+};
diff --git a/arch/arm/configs/xilinx_zynq_defconfig b/arch/arm/configs/xilinx_zynq_defconfig
index fae12de24..72a66fc7e 100644
--- a/arch/arm/configs/xilinx_zynq_defconfig
+++ b/arch/arm/configs/xilinx_zynq_defconfig
@@ -1,6 +1,4 @@
 CONFIG_LOCALVERSION="-xilinx"
-CONFIG_MTD_UBI=y
-CONFIG_UBIFS_FS=y
 CONFIG_SYSVIPC=y
 CONFIG_NO_HZ=y
 CONFIG_HIGH_RES_TIMERS=y
@@ -58,6 +56,8 @@ CONFIG_PACKET=y
 CONFIG_UNIX=y
 CONFIG_INET=y
 CONFIG_IP_MULTICAST=y
+CONFIG_IP_ADVANCED_ROUTER=y
+CONFIG_IP_MULTIPLE_TABLES=y
 CONFIG_IP_PNP=y
 CONFIG_IP_PNP_DHCP=y
 CONFIG_IP_PNP_BOOTP=y
@@ -80,11 +80,11 @@ CONFIG_MTD_CFI=y
 CONFIG_MTD_CFI_AMDSTD=y
 CONFIG_MTD_PHYSMAP=y
 CONFIG_MTD_PHYSMAP_OF=y
+CONFIG_MTD_M25P80=y
 CONFIG_MTD_NAND=y
 CONFIG_MTD_NAND_PL35X=y
 CONFIG_MTD_SPI_NOR=y
-CONFIG_MTD_M25P80=y
-CONFIG_JFFS2_FS=y
+CONFIG_MTD_UBI=y
 CONFIG_OF_OVERLAY=y
 CONFIG_OF_CONFIGFS=y
 CONFIG_BLK_DEV_LOOP=y
@@ -97,6 +97,7 @@ CONFIG_SCSI=y
 CONFIG_BLK_DEV_SD=y
 CONFIG_CHR_DEV_SG=y
 CONFIG_NETDEVICES=y
+CONFIG_TUN=y
 CONFIG_MACB=y
 # CONFIG_NET_VENDOR_CIRRUS is not set
 # CONFIG_NET_VENDOR_FARADAY is not set
@@ -122,6 +123,12 @@ CONFIG_INPUT_EVDEV=y
 CONFIG_KEYBOARD_GPIO=y
 CONFIG_KEYBOARD_GPIO_POLLED=y
 # CONFIG_LEGACY_PTYS is not set
+CONFIG_SERIAL_8250=y
+CONFIG_SERIAL_8250_CONSOLE=y
+CONFIG_SERIAL_8250_NR_UARTS=1
+CONFIG_SERIAL_8250_RUNTIME_UARTS=1
+CONFIG_SERIAL_8250_EXTENDED=y
+CONFIG_SERIAL_OF_PLATFORM=y
 CONFIG_SERIAL_XILINX_PS_UART=y
 CONFIG_SERIAL_XILINX_PS_UART_CONSOLE=y
 # CONFIG_HW_RANDOM is not set
@@ -136,6 +143,8 @@ CONFIG_SPI_ZYNQ_QSPI=y
 CONFIG_GPIO_SYSFS=y
 CONFIG_GPIO_XILINX=y
 CONFIG_GPIO_ZYNQ=y
+CONFIG_GPIO_PCA953X=y
+CONFIG_GPIO_PCA953X_IRQ=y
 CONFIG_PMBUS=y
 CONFIG_SENSORS_UCD9000=y
 CONFIG_SENSORS_UCD9200=y
@@ -217,6 +226,7 @@ CONFIG_REMOTEPROC=y
 CONFIG_ZYNQ_REMOTEPROC=m
 CONFIG_MEMORY=y
 CONFIG_IIO=y
+CONFIG_INA2XX_ADC=y
 CONFIG_XILINX_XADC=y
 CONFIG_RAS=y
 CONFIG_FPGA=y
@@ -227,11 +237,16 @@ CONFIG_XILINX_PR_DECOUPLER=y
 CONFIG_EXT2_FS=y
 CONFIG_EXT3_FS=y
 # CONFIG_DNOTIFY is not set
+CONFIG_OVERLAY_FS=y
 CONFIG_MSDOS_FS=y
 CONFIG_VFAT_FS=y
 CONFIG_TMPFS=y
 CONFIG_JFFS2_FS=y
 CONFIG_JFFS2_SUMMARY=y
+CONFIG_UBIFS_FS=y
+CONFIG_SQUASHFS=y
+# CONFIG_SQUASHFS_ZLIB is not set
+CONFIG_SQUASHFS_ZSTD=y
 CONFIG_NFS_FS=y
 CONFIG_ROOT_NFS=y
 CONFIG_NLS_CODEPAGE_437=y
diff --git a/drivers/net/phy/phy_device.c b/drivers/net/phy/phy_device.c
index 67f25ac29..dfee77997 100644
--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -1904,7 +1904,7 @@ static struct phy_driver genphy_driver = {
 	.name		= "Generic PHY",
 	.soft_reset	= genphy_no_soft_reset,
 	.config_init	= genphy_config_init,
-	.features	= PHY_GBIT_FEATURES | SUPPORTED_MII |
+	.features	= PHY_BASIC_FEATURES | SUPPORTED_MII |
 			  SUPPORTED_AUI | SUPPORTED_FIBRE |
 			  SUPPORTED_BNC,
 	.config_aneg	= genphy_config_aneg,
diff --git a/drivers/uio/uio_pdrv_genirq.c b/drivers/uio/uio_pdrv_genirq.c
index f598ecddc..6a13b8502 100644
--- a/drivers/uio/uio_pdrv_genirq.c
+++ b/drivers/uio/uio_pdrv_genirq.c
@@ -253,6 +253,7 @@ static const struct dev_pm_ops uio_pdrv_genirq_dev_pm_ops = {
 
 #ifdef CONFIG_OF
 static struct of_device_id uio_of_genirq_match[] = {
+        { .compatible = "generic-uio", },
 	{ /* This is filled with module_parm */ },
 	{ /* Sentinel */ },
 };
