diff --git a/include/configs/tsat_3500.h b/include/configs/tsat_3500.h
new file mode 100644
index 0000000..8b221db
--- /dev/null
+++ b/include/configs/tsat_3500.h
@@ -0,0 +1,84 @@
+/*
+ * (C) Copyright 2013 Xilinx, Inc.
+ * (C) Copyright 2015 Antmicro Ltd
+ * (C) Copyright 2020 TSAT AS
+ *
+ * Configuration for Enclustra Mars ZX3 SOM on TSAT3500 Digital Board
+ *
+ * See zynq-common.h for Zynq common configs
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#ifndef __CONFIG_TSAT_3500_H
+#define __CONFIG_TSAT_3500_H
+
+#define ENCLUSTRA_MARS_ZX
+#define CONFIG_CPU_FREQ_HZ		666666666
+
+#include <configs/zynq-common.h>
+#include <configs/enclustra_zx_common.h>
+
+#ifdef CONFIG_BOOTCOMMAND
+#undef CONFIG_BOOTCOMMAND
+#endif
+
+#ifdef RELEASE
+#define CONFIG_BOOTCOMMAND        "run tsat_rsa_verify && run qspiboot"
+#else
+#define CONFIG_BOOTCOMMAND        "run qspiboot"
+#endif
+
+#ifdef CONFIG_EXTRA_ENV_SETTINGS
+#undef CONFIG_EXTRA_ENV_SETTINGS
+#endif
+#define CONFIG_EXTRA_ENV_SETTINGS               \
+    "fdt_high=0x20000000\0"                     \
+    "initrd_high=0x20000000\0"                  \
+    "env_size=" ENVIRONMENT_SIZE "\0"           \
+    "boot_image_offset=0x40000\0"               \
+    "boot_image_size=0x500000\0"                \
+                                                \
+    "qspiboot="				 \
+        "sf probe;"                             \
+        "run tsat_qspiboot\0"                   \
+                                                \
+    "mmcdev=0\0"                                \
+                                                \
+    "sdboot="                                   \
+        "setenv mmcdev 0;"                      \
+        "run tsat_mmcboot\0"                    \
+                                                \
+    "tsat_firmware_selector=A\0"                    \
+    "tsat_def_args=console=ttyPS0,115200 ro earlyprintk\0"\
+    "tsat_bootscript_loadaddr=0x1000000\0"          \
+    "tsat_kernel_loadaddr=0x6000000\0"              \
+    "tsat_devicetree_loadaddr=0x6600000\0"          \
+                                                    \
+    "tsat_mmc_kernel_partition_a=5\0"               \
+    "tsat_mmc_kernel_partition_b=6\0"               \
+    "tsat_mmc_rootfs_partition_a=7\0"               \
+    "tsat_mmc_rootfs_partition_b=8\0"               \
+    "tsat_mmc_bootscript_image=u-boot-script.itb\0" \
+    "tsat_mmc_devicetree_image=linux.dtb\0"         \
+    "tsat_mmc_kernel_image=kernel-ramdisk-dtb.itb\0"\
+                                                    \
+    "tsat_mmcboot="                                 \
+        "mmc rescan && "                            \
+        "load mmc ${mmcdev} ${tsat_bootscript_loadaddr} ${tsat_mmc_bootscript_image} && "\
+        "source ${tsat_bootscript_loadaddr}:script-1\0"\
+                                                    \
+    "tsat_qspi_bootscript_offset=0x0500000\0"       \
+    "tsat_qspi_bootscript_size=0x40000\0"           \
+    "tsat_qspi_kernel_offset_a=0x0600000\0"         \
+    "tsat_qspi_kernel_offset_b=0x1200000\0"         \
+    "tsat_qspi_kernel_size=0xC00000\0"              \
+                                                    \
+    "tsat_qspiboot=echo Bootinq on QSPI Flash ...; "\
+        "zx_set_storage QSPI && "                   \
+        "sf read ${tsat_bootscript_loadaddr} ${tsat_qspi_bootscript_offset} ${tsat_qspi_bootscript_size} && "\
+        "source ${tsat_bootscript_loadaddr}:script-1\0"\
+    "fdtcontroladdr=0x2000000\0"                    \
+    "tsat_rsa_verify=sf probe && sf read 0x100000 ${boot_image_offset} ${boot_image_size} && zynqrsa 0x100000\0"
+
+#endif /* __CONFIG_TSAT_3500_H */
