diff --git a/board/xilinx/zynq/board.c b/board/xilinx/zynq/board.c
index f200737963..085f3b3706 100644
--- a/board/xilinx/zynq/board.c
+++ b/board/xilinx/zynq/board.c
@@ -347,7 +347,7 @@ int board_late_init(void)
 	uint32_t flash_size;
 
 	/* Probe the QSPI flash */
-	env_flash = spi_flash_probe(0, 0, 1000000, SPI_MODE_3);
+	env_flash = spi_flash_probe(0, 0, 0, SPI_MODE_3);
 
 	if (env_flash) {
 		/* Calculate the size in megabytes */
